src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,280 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
                     wfs[pos, :, :] = wf[:, sparsity_mask[unit_ind]]
 
 
+def extract_waveforms_to_single_buffer(
+    recording,
+    spikes,
+    unit_ids,
+    nbefore,
+    nafter,
+    mode="memmap",
+    return_scaled=False,
+    file_path=None,
+    dtype=None,
+    sparsity_mask=None,
+    copy=False,
+    job_name=None,
+    **job_kwargs,
+):
+    """
+    Allocate a single buffer (memmap or or shared memory) and then distribute every waveform into it.
+
+    Contrary to extract_waveforms_to_buffers() all waveforms are extracted in the same buffer, so the spike vector is
+    needed to recover waveforms unit by unit. Importantly in case of sparsity, the channels are not aligned across
+    units.
+
+    Important note: for the "shared_memory" mode wf_array_info contains reference to
+    the shared memmory buffer, this variable must be reference as long as arrays as used.
+    And this variable is also returned.
+    To avoid this a copy to non shared memmory can be perform at the end.
+
+    Parameters
+    ----------
+    recording: recording
+        The recording object
+    spikes: 1d numpy array with several fields
+        Spikes handled as a unique vector.
+        This vector can be obtained with: `spikes = Sorting.to_spike_vector()`
+    unit_ids: list ot numpy
+        List of unit_ids
+    nbefore: int
+        N samples before spike
+    nafter: int
+        N samples after spike
+    mode: str
+        Mode to use ('memmap' | 'shared_memory')
+    return_scaled: bool
+        Scale traces before exporting to buffer or not.
+    file_path: str or path
+        In case of memmap mode, file to save npy file.
+    dtype: numpy.dtype
+        dtype for waveforms buffer
+    sparsity_mask: None or array of bool
+        If not None shape must be must be (len(unit_ids), len(channel_ids))
+    copy: bool
+        If True (default), the output shared memory object is copied to a numpy standard array.
+        If copy=False then arrays_info is also return. Please keep in mind that arrays_info
+        need to be referenced as long as waveforms_by_units will be used otherwise it will be very hard to debug.
+        Also when copy=False the SharedMemory will need to be unlink manually
```

Comment:
What's the use case for offering this as an option. If there is, maybe you can mention here in the docstring. If there is None, maybe we can eliminate this parameter.

Reply:
Sharemmeory object is a dangerous object.
If you use a view if this one while your detroy the original memory you have very hard to debug stuff. (like segfault)
So a copy to a non shared mem buffer (standard numpy) is wise by default.

But in some cases like :  1. gather waveforms 2. make an average (for template) 3. remove the waveforms. Then in that case we can have this copy=False to avoid a new malloc in step 1.
This is more for internal use. (and a bit speculative at the moement, I admit)
The function that pierre need to rewritte should use this copy=False to be faster for instance.
look in the discussion.

Reply:
Thanks for the explanation, Sam.

I think that you should only return the shared memroy object (`shm`). No need to return the whole `arrays_info` dict.

Here my suggestion including what you mentioned above:

```
        If True (default), the output shared memory object is copied to a numpy standard array and no reference to the internal shared memory object is kept.
        If copy=False then the shared memory object is also returned. Please keep in mind that the shared memory object
        need to be referenced as long as waveforms_by_units will be used otherwise it might produce segmentation faults which are hard to debug.
        Also when copy=False the SharedMemory will need to be unlink manually if proper cleanup of resources is desired.
```

Reply:
Thank you.
I added your changes.
I think it is good to return the dict to be consistent with the behavior already established of `extract_waveforms_to_buffers()`

Reply:
Makes sense.